ostd/src/task/scheduler/fifo_scheduler.rs

```diff
@@ -0,0 +1,152 @@
+// SPDX-License-Identifier: MPL-2.0
+
+use alloc::{collections::VecDeque, sync::Arc, vec::Vec};
+
+use super::{EnqueueFlags, LocalRunQueue, Scheduler, UpdateFlags};
+use crate::{
+    cpu::{num_cpus, this_cpu},
+    sync::{SpinLock, SpinLockGuard},
+    task::Task,
+};
+
+/// A simple FIFO (First-In-First-Out) task scheduler.
+pub struct FifoScheduler<T: FifoSchedInfo> {
+    /// A thread-safe queue to hold tasks waiting to be executed.
+    rq: Vec<SpinLock<FifoRunQueue<T>>>,
+}
+
+impl<T: FifoSchedInfo> FifoScheduler<T> {
+    /// Creates a new instance of `FifoScheduler`.
+    fn new(ncpu: u32) -> Self {
+        let mut rq = Vec::new();
+        for _ in 0..ncpu {
+            rq.push(SpinLock::new(FifoRunQueue::new()));
+        }
+        Self { rq }
+    }
+
+    fn select_cpu(&self) -> u32 {
+        // FIXME: adopt more reasonable policy once we fully enable SMP.
+        0
+    }
+}
+
+impl<T: FifoSchedInfo + Send + Sync> Scheduler<T> for FifoScheduler<T> {
+    fn enqueue(&self, runnable: Arc<T>, flags: super::EnqueueFlags) -> Option<u32> {
+        let cpu = runnable.cpu();
+        if cpu.is_some() && flags == EnqueueFlags::Spawn {
+            return None;
+        }
+        let target_cpu = if cpu.is_some() {
+            cpu.unwrap()
+        } else {
+            self.select_cpu()
+        };
+        drop(cpu);
+
+        let mut rq = self.rq[target_cpu as usize].lock_irq_disabled();
+        if runnable.cpu().is_some() {
+            return None;
+        }
+        runnable.put_on_cpu(target_cpu);
+        rq.queue.push_back(FifoSchedEntity::new(runnable));
+
+        Some(target_cpu)
+    }
+
+    fn local_rq_with(&self, f: &mut dyn FnMut(&dyn LocalRunQueue<T>)) {
+        let local_rq: &FifoRunQueue<T> = &self.rq[this_cpu() as usize].lock_irq_disabled();
+        f(local_rq);
+    }
+
+    fn local_mut_rq_with(&self, f: &mut dyn FnMut(&mut dyn LocalRunQueue<T>)) {
+        let local_rq: &mut FifoRunQueue<T> = &mut self.rq[this_cpu() as usize].lock_irq_disabled();
+        f(local_rq);
+    }
+}
+
+struct FifoRunQueue<T: FifoSchedInfo> {
+    current: Option<FifoSchedEntity<T>>,
+    queue: VecDeque<FifoSchedEntity<T>>,
+}
+
+impl<T: FifoSchedInfo> FifoRunQueue<T> {
+    pub const fn new() -> Self {
+        Self {
+            current: None,
+            queue: VecDeque::new(),
+        }
+    }
+}
+
+impl<T: FifoSchedInfo> LocalRunQueue<T> for FifoRunQueue<T> {
+    fn current(&self) -> Option<&Arc<T>> {
+        self.current.as_ref().map(|entity| &entity.runnable)
+    }
+
+    fn update_current(&mut self, flags: super::UpdateFlags) -> bool {
+        !matches!(flags, UpdateFlags::Tick)
+    }
+
+    fn pick_next_current(&mut self) -> Option<Arc<T>> {
+        let prev = self.current.take();
```

Comment:
Again, do not `take` first then put it back. Call `current.replace(...)` to combine the two operations of setting the new current and taking the old one in one call.